runtime/service.go

```diff
@@ -1148,54 +1148,61 @@ func (s *service) Shutdown(requestCtx context.Context, req *taskAPI.ShutdownRequ
 		return &ptypes.Empty{}, nil
 	}
 
-	shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), defaultShutdownTimeout*time.Second)
-	defer shutdownCancel()
-
-	if err := s.shutdown(requestCtx, shutdownCtx, req); err != nil {
+	if err := s.shutdown(requestCtx, defaultShutdownTimeout, req); err != nil {
 		return &ptypes.Empty{}, err
 	}
 
 	return &ptypes.Empty{}, nil
 }
 
-// shutdown will shutdown the VMM. requestCtx is used to issue requests and shutdownCtx is used for waiting a graceful shutdown procedure.
-// When shutdownCtx's Done channel is closed during the graceful shutdown procedure, the VMM is forcefully killed.
-func (s *service) shutdown(requestCtx, shutdownCtx context.Context, req *taskAPI.ShutdownRequest) error {
-	// cancel the shim at the end, in case even the last StopVMM() doesn't work
-	defer s.shimCancel()
-
-	s.logger.Info("stopping the VM")
-	if _, err := s.agentClient.Shutdown(requestCtx, req); err != nil {
-		return err
-	}
-
-	shutdownCh := make(chan error)
+// shutdown will stop the VMM within the provided timeout. It attempts to shutdown gracefully by having
+// agent stop (which is presumed to cause the VM to begin a reboot) and then waiting for the VMM process
+// to exit (via the s.shimCtx.Done() channel). If that fails, StopVMM will be called to force a shutdown (currently
+// via sending SIGTERM). If that fails, the VMM will still be killed via SIGKILL when the shimCtx is canceled.
+func (s *service) shutdown(
+	requestCtx context.Context,
```

Comment:
This request context isn't used anywhere but creating the timeout context. How do you feel about passing that context instead of the `requestCtx`?

Reply:
All the callers care about is that `shutdown` should have a timeout associated with it. If we have the callers create contexts, there's more duplicated code, more parameters to pass to `shutdown` (both a context and a cancel func) and more places you need to look to understand how the `shutdown` code works. I think it's simpler to just keep the fact that the timeout is currently enforced with a derived context inside `shutdown` rather than expose more to the callers.

Reply:
That's true with the `requestCtx` as well. You need to track what context is actually being passed in, so I don't think it is too much more to ask for. However, I don't feel too strongly on this.